### PR TITLE
Use joblib hashing to generate cache key to ensure repeatability

### PR DESCRIPTION
**Pull request by wild-endeavor**
# TL;DR
Thanks to @bethebunny for finding/fixing this.

When the type of a literal is a `dict` or `Dict` (i.e. no type specified for the value), the type engine converts it to a protobuf struct.  If the protobuf struct contains more than one field, then the cache key generated under the old code which ran base64 encoded a cloudpickle dump would sometimes change between calls, which led to local caching not working.

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [x] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/<number>

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

### CODE REVIEW

**Comment by wild-endeavor** [comment]
nm ignore this for now.  need to add more to this.

**Comment by wild-endeavor** [comment]
@eapolinario can you take a look at this when you get a chance please?  I don't understand the unit tests that are failing and can't find any references to explain what's happening.  Why do we not want ints to cache?

**Review by eapolinario — APPROVED** [review]
